src/ast/seq_decl_plugin.cpp

```diff
@@ -1463,3 +1463,191 @@ bool seq_util::re::is_epsilon(expr* r) const {
 app* seq_util::re::mk_epsilon(sort* seq_sort) {
     return mk_to_re(u.str.mk_empty(seq_sort));
 }
+
+/*
+  Provides a standard pretty printed view of the regex r when possible.
+*/
+std::string seq_util::re::pp(expr* r) {
+    SASSERT(u.is_re(r));
+    std::ostringstream buffer;
+    pp_compact_to_buffer(buffer, r);
+    return buffer.str();
+}
+
+void seq_util::re::pp_compact_to_buffer(std::ostringstream& buffer, expr* r) {
+    SASSERT(u.is_re(r));
+    expr* r1 = nullptr, * r2 = nullptr, * s = nullptr, * s2 = nullptr;
+    unsigned lo = 0, hi = 0;
+    if (is_full_char(r))
+        buffer << ".";
+    else if (is_full_seq(r))
+        buffer << ".*";
+    else if (is_to_re(r, s))
+        pp_compact_helper_seq(buffer, s);
+    else if (is_range(r, s, s2))
+        pp_compact_helper_range(buffer, s, s2);
+    else if (is_epsilon(r))
+        buffer << "()";
+    else if (is_empty(r))
+        buffer << "[]";
+    else if (is_concat(r, r1, r2)) {
+        pp_compact_to_buffer(buffer, r1);
+        pp_compact_to_buffer(buffer, r2);
+    }
+    else if (is_union(r, r1, r2)) {
+        pp_compact_to_buffer(buffer, r1);
+        buffer << "|";
+        pp_compact_to_buffer(buffer, r2);
+    }
+    else if (is_intersection(r, r1, r2)) {
+        buffer << "(";
+        pp_compact_to_buffer(buffer, r1);
+        buffer << ")&(";
+        pp_compact_to_buffer(buffer, r2);
+        buffer << ")";
+    }
+    else if (is_complement(r, r1)) {
+        buffer << "~";
+        if (pp_can_skip_parenth(r1))
+            pp_compact_to_buffer(buffer, r1);
+        else {
+            buffer << "(";
+            pp_compact_to_buffer(buffer, r1);
+            buffer << ")";
+        }
+    }
+    else if (is_plus(r, r1))
+        if (pp_can_skip_parenth(r1)) {
+            pp_compact_to_buffer(buffer, r1);
+            buffer << "+";
+        }
+        else {
+            buffer << "(";
+            pp_compact_to_buffer(buffer, r1);
+            buffer << ")+";
+        }
+    else if (is_star(r, r1))
+        if (pp_can_skip_parenth(r1)) {
+            pp_compact_to_buffer(buffer, r1);
+            buffer << "*";
+        }
+        else {
+            buffer << "(";
+            pp_compact_to_buffer(buffer, r1);
+            buffer << ")*";
+        }
+    else if (is_loop(r, r1, lo))
+        if (pp_can_skip_parenth(r1))
+        {
+            pp_compact_to_buffer(buffer, r1);
+            buffer << "{" << std::to_string(lo) << ",}";
+        }
+        else {
+            buffer << "(";
+            pp_compact_to_buffer(buffer, r1);
+            buffer << "){" << std::to_string(lo) << ",}";
```

Comment:
I think there is already printers to std::ostream& so no need for std::to_string after you change the parameter type

Also, rename buffer to "out"

Reply:
ok